File: navground_examples_py/navground_examples_py/my_behavior.py
from __future__ import annotations

from typing import SupportsFloat

import numpy as np
from navground import core


class PyIdleBehavior(core.Behavior, name="PyIdle"):

    def __init__(self,
                 kinematics: core.Kinematics | None = None,
                 radius: float = 0) -> None:
        super().__init__(kinematics, radius)
        self._ignore_obstacles = False

    def desired_velocity_towards_point(
            self, point: core.Vector2Like, speed: SupportsFloat,
            time_step: SupportsFloat) -> core.Vector2:
        if self.ignore_obstacles:
            return core.clamp_norm(np.asarray(point) - self.position, speed)
        return core.zeros2()

    def desired_velocity_towards_velocity(
            self, velocity: core.Vector2Like,
            time_step: SupportsFloat) -> core.Vector2:
        if self.ignore_obstacles:
            return np.asarray(velocity)
        return core.zeros2()

    @property
    @core.register(False,
                   "whether to move towards the target, ignoring obstacles")
    def ignore_obstacles(self) -> bool:
        return self._ignore_obstacles

    @ignore_obstacles.setter
    def ignore_obstacles(self, value: bool) -> None:
        self._ignore_obstacles = value

File: navground_examples/src/behavior/my_behavior.cpp
/**
 * @author Jerome Guzzi - <jerome@idsia.ch>
 */

#include "./my_behavior.h"

namespace navground::core {

const std::string IdleBehavior::type = register_type<IdleBehavior>(
    "Idle", {{"ignore_obstacles",
              Property::make(
                  &IdleBehavior::get_ignore_obstacles,
                  &IdleBehavior::set_ignore_obstacles, false,
                  "whether to move towards the target, ignoring obstacles")}});

} // namespace navground::core
